File: python/mergesort.py

def merge(a, b):
	if len(a)*len(b) == 0:
		return a+b

	v = (a[0] < b[0] and a or b).pop(0)
	return [v] + merge(a, b)

def mergesort(lst):
	if len(lst) < 2:
		return lst

	m = len(lst)/2
	return merge(mergesort(lst[:m]), mergesort(lst[m:]))

mlst = [10, 9, 8, 4, 5, 6, 7, 3, 2, 1]
sorted = mergesort(mlst)
print sorted

File: c++/mergesort.cpp

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

template<typename IT> void merge(IT begin, IT middle, IT end, IT res)
{
	IT a = begin, b = middle, r = res;

	while (a < middle && b < end)
		if (*a < *b) *r++ = *a++;
		else *r++ = *b++;

	while (a < middle) *r++ = *a++;
	while (b < end) *r++ = *b++;
	while (begin < end) *begin++ = *res++;
}

template<typename IT> void mergesort(IT begin, IT end, IT res)
{
	int s = end-begin;
	if (s > 1)
	{
		IT middle = begin+s/2;
		mergesort(begin, middle, res);
		mergesort(middle, end, res);
		merge(begin, middle, end, res);
	}
}

int main()
{
	int lst[] = {10, 9, 8, 4, 5, 6, 7, 3, 2, 1};
    int sorted[10];

	mergesort(lst, lst + 10, sorted);

	for (int i = 0; i < 10; i++)
		cout << sorted[i] << " ";
	cout << endl;
	return 0;
}
